numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1571,7 +1571,330 @@ string_expandtabs_strided_loop(PyArrayMethod_Context *context,
+        PyArray_Descr *loop_descrs[], npy_intp *NPY_UNUSED(view_offset))
+{
+    PyArray_StringDTypeObject *input_descr = (PyArray_StringDTypeObject *)given_descrs[0];
+    PyArray_StringDTypeObject *fill_descr = (PyArray_StringDTypeObject *)given_descrs[2];
+
+    int eq_res = _eq_comparison(input_descr->coerce, fill_descr->coerce,
+                                input_descr->na_object, fill_descr->na_object);
+
+    if (eq_res < 0) {
+        return (NPY_CASTING)-1;
+    }
+
+    if (eq_res != 1) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Can only text justification operations with equal"
+                        "StringDType instances.");
+        return (NPY_CASTING)-1;
+    }
+
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+    Py_INCREF(given_descrs[1]);
+    loop_descrs[1] = given_descrs[1];
+    Py_INCREF(given_descrs[2]);
+    loop_descrs[2] = given_descrs[2];
+
+    PyArray_Descr *out_descr = NULL;
+
+    if (given_descrs[3] == NULL) {
+        out_descr = (PyArray_Descr *)new_stringdtype_instance(
+                ((PyArray_StringDTypeObject *)given_descrs[1])->na_object,
+                ((PyArray_StringDTypeObject *)given_descrs[1])->coerce);
+
+        if (out_descr == NULL) {
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        Py_INCREF(given_descrs[3]);
+        out_descr = given_descrs[3];
+    }
+
+    loop_descrs[3] = out_descr;
+
+    return NPY_NO_CASTING;
+}
+
+
+static const char* CENTER_NAME = "center";
+static const char* LJUST_NAME = "ljust";
+static const char* RJUST_NAME = "rjust";
+
+template <ENCODING enc>
+static int
+center_ljust_rjust_strided_loop(PyArrayMethod_Context *context,
+                                char *const data[],
+                                npy_intp const dimensions[],
+                                npy_intp const strides[],
+                                NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_StringDTypeObject *s1descr = (PyArray_StringDTypeObject *)context->descriptors[0];
+    int has_null = s1descr->na_object != NULL;
+    int has_nan_na = s1descr->has_nan_na;
+    int has_string_na = s1descr->has_string_na;
+    const npy_static_string *default_string = &s1descr->default_string;
+    npy_intp N = dimensions[0];
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *in3 = data[2];
+    char *out = data[3];
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp in3_stride = strides[2];
+    npy_intp out_stride = strides[3];
+
+    npy_string_allocator *allocators[4] = {};
+    NpyString_acquire_allocators(4, context->descriptors, allocators);
+    npy_string_allocator *s1allocator = allocators[0];
+    // allocators[1] is NULL
+    npy_string_allocator *s2allocator = allocators[2];
+    npy_string_allocator *oallocator = allocators[3];
+
+    JUSTPOSITION pos = *(JUSTPOSITION *)(context->method->static_data);
+    const char* ufunc_name = NULL;
```

Comment:
Why not just `ufunc_name  = ((PyUFuncObject *)context->caller)->name`, as done elsewhere?

Reply:
Oops, missed that from your refactors